### PR TITLE
Qt: remove macOS launch-at-startup when compiled with > macOS 10.11, fix memory missmanagement

### LINKED ISSUE #15142: gui: crash on macOS when unchecking "start Bitcoin Core on system login"
# Describe the issue
Unchecking the "Start Bitcoin Core on System login" causes 0.17.1 to crash on Mac OS.
# What behavior did you expect?
To continue running and not start on reboot.
# What was the actual behavior (provide screenshots if the issue is GUI-related)? 
See video: 
https://twitter.com/_drgo/status/1083409084655665152?s=21
# How reliably can you reproduce the issue, what are the steps to do so? 
Reliably. See above.
# What version of Bitcoin Core are you using, where did you get it (website, self-compiled, etc)? 
0.17.1; downloaded from link on Bitcoin.org; shasum -a 256 matched.
# What type of machine are you observing the error on (OS/CPU and disk type)? 
Mac Mini 2.6 ghz dual core i5, Mac OS Mojave 10.14.2

# Debug.log 
Doesn't seem to informative...

https://pastebin.com/qckAe3J1

### PR DESCRIPTION
**Pull request by jonasschnelli**
The launch-at-startup API Bitcoin Core uses on macOS where removed in macOS 10.11 leading to a segmentation-fault due to the weak-linking when not actively compiled against SDK 10.11 (`-mmacosx-version-min=10.11`)

This PR removes the launch-at-startup feature on macOS when compiled with macOS min version > 10.11 (the default is always the macOS version you compile on).

**The depends built binaries (Gitian) are not affected since we are building with min macOS 10.10.**

Users self compiling on macOS > 10.11 can re-enable the feature by compiling with min version <= 10.11 (`CXXFLAGS="-mmacosx-version-min=10.11" CFLAGS="-mmacosx-version-min=10.11" ./configure`)

**Isn't there a new API from Apple?**
Yes, [there is](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLoginItems.html).
It will require to create a helper application which needs to be embedded in the .app folder (needs code signing as well). Developers willing to go down that rabbit hole are welcome.

Fixes #15142

### CODE REVIEW

**Comment by hebasto** [comment]
utACK 8441c90af88fb0f08762cf56e508803ed4b47d4c

File: doc/release-notes.md [hunk 1]
```diff
@@ -262,6 +262,11 @@ Graphical User Interface (GUI)
   balance shown if the wallet was created using the `createwallet` RPC
   and the `disable_private_keys` parameter was set to true.
 
+- The launch-on-startup option in no longer available on macOS if
```
Comment (hebasto) [inline doc/release-notes.md]: `in` --> `is` ?

**Comment by fanquake** [comment]
Compiling 8441c90 on macOS `10.14.2`:
```
./autogen.sh && ./configure && make check -j6
```
The option is disabled.
pr - no 10 11

However, when passing `-mmacosx-version-min=10.11` to  `./configure`:
```
./autogen.sh
CXXFLAGS="-mmacosx-version-min=10.11" CFLAGS="-mmacosx-version-min=10.11" ./configure
make check -j6
lldb src/qt/bitcoin-qt
```
I'm still seeing the same segfault behaviour as #15142:
```
Current executable set to 'src/qt/bitcoin-qt' (x86_64).
(lldb) run
Process 20678 launched: '/Users/michael/github/bitcoin/src/qt/bitcoin-qt' (x86_64)
2019-01-20 11:09:45.261118+0800 bitcoin-qt[20678:2180338] MessageTracer: Falling back to default whitelist
Process 20678 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x20)
    frame #0: 0x00007fff72cb58bf libobjc.A.dylib`objc_retain + 31
libobjc.A.dylib`objc_retain:
->  0x7fff72cb58bf <+31>: testb  $0x4, 0x20(%rcx)
    0x7fff72cb58c3 <+35>: je     0x7fff72cb58f6            ; <+86>
    0x7fff72cb58c5 <+37>: testb  $0x1, %al
    0x7fff72cb58c7 <+39>: je     0x7fff72cb58e9            ; <+73>
Target 0: (bitcoin-qt) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x20)
  * frame #0: 0x00007fff72cb58bf libobjc.A.dylib`objc_retain + 31
    frame #1: 0x00007fff484b8f18 SharedFileList`-[SFLLoginItemList removeItem:error:] + 35
    frame #2: 0x00007fff484bf3c7 SharedFileList`LSSharedFileListItemRemove + 127
    frame #3: 0x000000010007041f bitcoin-qt`GUIUtil::SetStartOnSystemStartup(bool) + 223
    frame #4: 0x00000001000a25fa bitcoin-qt`OptionsModel::setData(QModelIndex const&, QVariant const&, int) + 202
    frame #5: 0x000000010241a7a7 QtWidgets`QItemDelegate::setModelData(QWidget*, QAbstractItemModel*, QModelIndex const&) const + 295
    frame #6: 0x0000000102423b48 QtWidgets`___lldb_unnamed_symbol4564$$QtWidgets + 248
    frame #7: 0x0000000102424c7b QtWidgets`QDataWidgetMapper::submit() + 59
    frame #8: 0x000000010009573b bitcoin-qt`OptionsDialog::on_okButton_clicked() + 43
    frame #9: 0x00000001001e0f9a bitcoin-qt`OptionsDialog::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) + 202
    frame #10: 0x00000001001e136b bitcoin-qt`OptionsDialog::qt_metacall(QMetaObject::Call, int, void**) + 139
    frame #11: 0x0000000102e79bbd QtCore`QMetaObject::activate(QObject*, int, int, void**) + 1773
    frame #12: 0x000000010227aacf QtWidgets`___lldb_unnamed_symbol1034$$QtWidgets + 111
    frame #13: 0x000000010227a96c QtWidgets`___lldb_unnamed_symbol1032$$QtWidgets + 236
    frame #14: 0x000000010227ba9f QtWidgets`QAbstractButton::mouseReleaseEvent(QMouseEvent*) + 271
```

**Comment by jonasschnelli** [comment]
@fanquake: your right. There is something wrong with out implementation... I'll fix that soon.

**Comment by jonasschnelli** [comment]
Current macOS launch at login code does release an item which belongs to an array that gets release later (double release) which causes a crash on certain macOS versions.

I added a commit that fixes the memory mismanagement.

File: doc/release-notes.md [hunk 2]
```diff
@@ -262,6 +262,11 @@ Graphical User Interface (GUI)
   balance shown if the wallet was created using the `createwallet` RPC
   and the `disable_private_keys` parameter was set to true.
 
+- The launch-on-startup option is no longer available on macOS if
+  compiled with macosx min version greater then 10.11 (use
+  CXXFLAGS="-mmacosx-version-min=10.11" for setting the deployment
```
Comment (hebasto) [inline doc/release-notes.md]: `CFLAGS="-mmacosx-version-min=10.11"` could be added as well.
Without `CFLAGS` there are linker warnings:
screenshot from 2019-01-21 11-38-04

File: src/qt/guiutil.cpp [hunk 3]
```diff
@@ -683,13 +683,12 @@ bool SetStartOnSystemStartup(bool fAutoStart)
 }
 
 
-#elif defined(Q_OS_MAC)
+#elif defined(Q_OS_MAC) && defined(MAC_OS_X_VERSION_MIN_REQUIRED) && MAC_OS_X_VERSION_MIN_REQUIRED <= 101100
 // based on: https://github.com/Mozketo/LaunchAtLoginController/blob/master/LaunchAtLoginController.m
 
-LSSharedFileListItemRef findStartupItemInList(LSSharedFileListRef list, CFURLRef findUrl);
-LSSharedFileListItemRef findStartupItemInList(LSSharedFileListRef list, CFURLRef findUrl)
+LSSharedFileListItemRef findStartupItemInList(CFArrayRef listSnapshot, LSSharedFileListRef list, CFURLRef findUrl);
```
Comment (hebasto) [inline src/qt/guiutil.cpp]: This function declaration seems unnecessary. Could it be removed?

File: src/qt/guiutil.cpp [hunk 4]
```diff
@@ -734,10 +730,12 @@ bool GetStartOnSystemStartup()
...
     LSSharedFileListRef loginItems = LSSharedFileListCreate(nullptr, kLSSharedFileListSessionLoginItems, nullptr);
-    LSSharedFileListItemRef foundItem = findStartupItemInList(loginItems, bitcoinAppUrl);
-
+    CFArrayRef listSnapshot = LSSharedFileListCopySnapshot(loginItems, nullptr);
+    bool res = (findStartupItemInList(listSnapshot, loginItems, bitcoinAppUrl) != nullptr);
     CFRelease(bitcoinAppUrl);
-    return !!foundItem; // return boolified object
+    CFRelease(loginItems);
+    CFRelease(listSnapshot);
+    return res; // return boolified object
```
Comment (hebasto) [inline src/qt/guiutil.cpp]: `res` is a `bool` already. Could the comment be removed?

**Review by hebasto — CHANGES_REQUESTED** [review]
Tested 541d287b805ecaad1d9810b5a776d2e0d8b87233 on macOS 10.13.6.

**Comment by hebasto** [comment]
tACK 35f5cb791de6bd47b62e031a74d885f00efd2585 (macOS 10.13.6 + Qt 5.11.2)

> The depends built binaries (Gitian) are not affected since we are building with min macOS 10.10.

Note: [official v0.17.1](https://bitcoincore.org/bin/bitcoin-core-0.17.1/bitcoin-0.17.1-osx.dmg) _is_ affected.
Could backport?

File: doc/release-notes.md [hunk 5]
```diff
@@ -262,6 +262,12 @@ Graphical User Interface (GUI)
   balance shown if the wallet was created using the `createwallet` RPC
   and the `disable_private_keys` parameter was set to true.
 
+- The launch-on-startup option is no longer available on macOS if
+  compiled with macosx min version greater then 10.11 (use
```
Comment (practicalswift) [inline doc/release-notes.md]: Should be "greater than" :-)

Comment (jonasschnelli) [reply doc/release-notes.md]: Fixed.

**Comment by laanwj** [comment]
utACK da6011826a730837b59aef5664f3feab4787c9bc

**Pull request merged**